patches/flashtools.patch

```diff
@@ -0,0 +1,302 @@
++	if (argc <= 1)
++	{
++		fprintf(stderr, "%s", usage);
++		return EXIT_FAILURE;
++	}
++
++	int opt;
++	int do_read = 0;
++	int do_list = 0;
++	const char * filename = NULL;
++	while ((opt = getopt_long(argc, argv, "h?vlr:", long_options, NULL)) != -1)
++	{
++		switch(opt)
++		{
++		case 'v':
++			verbose++;
++			break;
++		case 'l':
++			do_list = 1;
++			break;
++		case 'r':
++			do_read = 1;
++			filename = optarg;
++			break;
++		case '?': case 'h':
++			printf("%s", usage);
++			return EXIT_SUCCESS;
++		default:
++			fprintf(stderr, "%s", usage);
++			return EXIT_FAILURE;
++		}
++	}
++
++	argc -= optind;
++	argv += optind;
++	if (argc != 0)
++	{
++		fprintf(stderr, "%s: Excess arguments?\n", prog_name);
++		return EXIT_FAILURE;
++	}
++
++	int fd;
++	fd = open("/dev/mem", O_RDONLY);
++	if (fd < 0) {
++		fprintf(stderr, "Failed to open /dev/mem : %s\n", strerror(errno));
++		return EXIT_FAILURE;
++	}
++
++	uint64_t end = 0x100000000;
++	if (verbose) {
++		printf("Seeking to %lx\n", end-4);
++	}
++	if (lseek(fd, end-4, SEEK_SET) == -1) {
++		fprintf(stderr, "Failed to seek to %lx\n", end-4);
++		return EXIT_FAILURE;
++	}
++
++	int32_t header_delta;
++	if (read(fd, &header_delta, sizeof(header_delta)) < 0) {
++		fprintf(stderr, "Failed to read header offset\n");
++		return EXIT_FAILURE;
++	}
++	if (verbose) {
++		printf("Header Offset: %d\n", header_delta);
++	}
++
++	uint64_t header_off;
++	if (header_delta < 0) {
++		header_off = end - ((uint64_t)(-header_delta));
++	} else {
++		header_off = ((uint64_t)header_delta);
++	}
++
++	if (verbose) {
++		printf("Seeking to %lx\n", header_off);
++	}
++	if (lseek(fd, header_off, SEEK_SET) == -1) {
++		fprintf(stderr, "Failed to seek to %lx\n", header_off);
++		return EXIT_FAILURE;
++	}
++
++	struct cbfs_header header;
++	if (read(fd, &header, sizeof(struct cbfs_header)) < 0) {
++		fprintf(stderr, "Failed to read header\n");
++		return EXIT_FAILURE;
++	}
++	header.magic = ntohl(header.magic);
++	header.version = ntohl(header.version);
++	header.romsize = ntohl(header.romsize);
++	header.bootblocksize = ntohl(header.bootblocksize);
++	header.align = ntohl(header.align);
++	header.offset = ntohl(header.offset);
++	header.architecture = ntohl(header.architecture);
++
++	if (verbose) {
++		printf("Header magic          : %x\n", header.magic);
++		printf("Header version        : %x\n", header.version);
++		printf("Header ROM size       : %x\n", header.romsize);
++		printf("Header boot block size: %x\n", header.bootblocksize);
++		printf("Header align          : %x\n", header.align);
++		printf("Header offset         : %x\n", header.offset);
++		printf("Header arch           : %x\n", header.architecture);
++	}
++
++	if (header.magic != CBFS_HEADER_MAGIC) {
++		fprintf(stderr, "Failed to find valid header\n");
++		return EXIT_FAILURE;
++	}
++
++	uint32_t align = header.align;
++	// loop through files
++	uint64_t off = end - ((uint64_t) header.romsize) + ((uint64_t) header.offset);
++	while (1) {
```

Comment:
Instead of an infinite loop (which breaks when it's outside the bounds of the cbfs), it would be better to do : 
`   while (off < end) { `